### PR TITLE
Align the SpikeSim dense energy consumption model and tighten the runtime scope

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- Refactor the `spikesim` energy estimation pipeline, by default align with the original SpikeSim dense PE-cycle accounting, and retain runtime stage/call discovery
- Tighten the scope to Conv2d forward inference, supplement IF/LIF neuron hypothesis validation, and avoid misclassifying incompatible neurons as SpikeSim hardware models
- Synchronize and reorganize `op_counter` related infrastructure and tests, covering SpikeSim default mode and explicit `event` extension mode

## Testing

- The file `test/activation_based/test_op_counter_spikesim_event_energy.py` passes
- The files `test/activation_based/test_synops.py`, `test/activation_based/test_op_counter_neuromc_runtime.py`, `test/activation_based/test_ac.py`, `test/activation_based/test_mac.py` pass

## Summary by CodeRabbit

* **New Features**
  * SpikeSimCounter exposed for runtime energy profiling.
  * Config options: activity_mode ("dense" default or "event") and require_if_lif_neurons toggle.

* **Improvements**
  * Kernel-aware per-PE/per-kernel-cycle energy and per-stage/component breakdowns.
  * Profiler collects and surfaces clearer warnings (training vs inference, unsupported convs, neuron support, experimental event mode); accepts injected warnings.
  * Dense mode reports dense PE-cycle counts regardless of input sparsity.

* **Tests**
  * Expanded tests for modes, validation, warnings, dense vs event accounting, and neuron support.

[Review Change Stack](https://app.coderabbit.ai/change-stack/fangwei123456/spikingjelly/pull/682?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by AllenYolk** [comment]
@coderabbitai review

File: spikingjelly/activation_based/op_counter/spikesim/core.py [hunk 1]
```diff
@@ -61,6 +89,9 @@ class SpikeSimEventEnergyReport:
     event_stats_by_stage: dict[str, dict[str, Any]]
     stage_metadata: dict[str, dict[str, Any]]
     warnings: list[str]
+    total_pj: float
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/core.py]: 💡 **Suggestion (Design):** `total_pj` stores the same value as `energy_total_pj`. Having both fields creates ambiguity for downstream consumers — which one is canonical? Consider either removing `total_pj` or deprecating `energy_total_pj` in favor of the new field.

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Approved ✅** (0 critical, 0 warnings, 2 suggestions)

**PR:** #682 — Align the SpikeSim dense energy consumption model and tighten the runtime scope
**Author:** @AllenYolk
**Files changed:** 7 (+352 -180)

### 💡 Suggestions

#### (Maintainability) Dead code: `_compute_spikesim_dense_stage_energy` — formulas.py:84

The private function `_compute_spikesim_dense_stage_energy` in `formulas.py` is no longer called from anywhere. It was replaced by `compute_spikesim_dense_energy_breakdown` (new) + `config.pe_cycle_energy_for_kernel_pj` (extracted into config.py). Since `_compute_spikesim_dense_stage_energy` is not in `__all__` and has no remaining callers, consider removing it to reduce maintenance surface.

#### (Design) Redundant report fields `total_pj` and `energy_total_pj` — core.py:92-93, 251-260

`SpikeSimEventEnergyReport` now has both `energy_total_pj` (existing field) and `total_pj` (new field, line 92-93), both populated from the same `totals_dict.get("total_pj", 0.0)` (lines 251, 260). This creates API ambiguity — downstream consumers reading `report.total_pj` vs `report.energy_total_pj` get the same value, but it's unclear which is canonical. Consider either:
- Deprecating `energy_total_pj` and keeping only `total_pj`, or
- Removing `total_pj` since `energy_total_pj` already exists

#### (Suggestion) Weak assertion in `test_spikesim_event_energy_counter_base_contract_nonempty` — test line 175

The assertion `records["Global"][torch.ops.aten.convolution.default] > 1` (line 175) is quite weak — the return value changed from always-1 to `_dense_pe_cycles(...)` = 64 for the test params. Either assert the exact expected value (`p_i * q_i * num_sites = ceil(3/4) * ceil(4/4) * 1*8*8 = 64`) or change the assertion pattern to a more descriptive check.

### ✅ Looks Good

- **Clean architecture refactor.** Replacing the dual-path (`ActiveModuleTracker` + `SpikeSimEventTraceMode`) with the unified `DispatchCounterMode` + `SpikeSimCounter` is a significant simplification. The counter now handles both event and dense modes through the same dispatch path.

- **Tightened scope validation.** The `_is_forward_inference_conv` guard correctly restricts profiling to `nn.Conv2d` modules in eval mode with grad disabled. The `training` check in `estimate_spikesim_event_energy` provides an additional safety net.

- **IF/LIF neuron hypothesis check.** `_unsupported_neuron_modules` + `require_if_lif_neurons` is a well-designed safety rail — warns/raises when the model contains incompatible neurons (e.g., `ParametricLIFNode`) that violate SpikeSim's IF/LIF assumption.

- **`pe_cycle_energy_for_kernel_pj` extraction.** Pulling the kernel-dependent PE cycle energy formula into `SpikeSimEnergyConfig` (config.py:56-65) eliminates a duplicate computation that existed in the old `_compute_spikesim_dense_stage_energy` and event breakdown code paths.

- **`is_binary_tensor` reuse.** Replacing the local `_is_spike_like` (now removed) with the shared `is_binary_tensor` from `op_counter.base` improves consistency across counters.

- **Defensive fallback in `compute_spikesim_event_energy_breakdown`.** Lines 33-34 correctly delegate to `compute_spikesim_dense_energy_breakdown` when `activity_mode != "event"`, ensuring direct callers of the event breakdown function get correct behavior regardless of config.

- **Comprehensive test coverage.** The new tests cover: dense mode sparsity independence (`test_spikesim_default_dense_mode_ignores_input_sparsity`), training mode warnings/raises (`test_spikesim_training_scope_warns_or_raises`), IF/LIF neuron acceptance (`test_spikesim_allows_if_and_lif_neurons`), unsupported neuron detection (`test_spikesim_unsupported_neuron_warns_or_raises`), and the disable path (`test_spikesim_unsupported_neuron_check_can_be_disabled`).

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review

File: spikingjelly/activation_based/op_counter/spikesim/config.py [hunk 2]
```diff
@@ -120,6 +126,26 @@ def patch_control_energy_pj(self) -> float:
             + (b / 8.0) * (b / 8.0) * (self.adc_pj + self.mux_pj)
         )
 
+    @property
+    def pe_cycle_energy_pj(self) -> float:
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/config.py]: **Suggestion:** The `pe_cycle_energy_pj` property (line 130) is defined but never used anywhere in the codebase or tests — not in `formulas.py`, not in `counter.py`, not in `core.py`. The energy pipeline uses `pe_cycle_energy_for_kernel_pj()` instead.

If this is kept as a convenience API for the common 3×3 kernel case, consider adding a brief docstring or at least a comment noting it's a 3×3-specific shorthand. If not intended as public API, it should be removed to avoid dead code.

Note: the test helper `_dense_stage_energy` in the test file also duplicates the `pe_cycle_energy_for_kernel_pj(…)` formula manually — it could reuse the config method.

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 3]
```diff
@@ -360,48 +381,37 @@ def _handle_convolution(self, scope: str, args: tuple[Any, ...], out) -> None:
...
+                "event"
+                if self.config.activity_mode == "event" and spike_like_input
+                else "dense"
+            )
             print(
-                f"SpikeSimEventCounter: {scope} - aten.convolution.default "
+                f"SpikeSimCounter: {scope} - aten.convolution.default "
                 f"[{mode}] x={tuple(x.shape)} w={tuple(w.shape)} out={tuple(out.shape)}"
             )
+        return self._dense_pe_cycles(w=w, out=out)
```
Comment (AllenYolk) [inline spikingjelly/activation_based/op_counter/spikesim/counter.py]: **Minor (Design):** `_dense_pe_cycles()` is called twice per convolution dispatch — once in `_update_stage` (line 303) for accumulating `stage_stats.dense_pe_cycle_count`, and once in `_handle_convolution` return (line 393) for the `count()` return value and `records[]` accumulation.

Both compute the same value from the same `w` and `out` arguments, so the result is correctly consistent (as verified by the `counter_base_contract_nonempty` test). However, this is unnecessary double work — the return value from `_handle_convolution` could reuse the already-computed `_dense_pe_cycles` result from within `_update_stage` by returning it, avoiding the redundant call.

Not blocking, but worth cleaning up if there's a next refactoring pass.

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: Comment** (2 suggestions, no blocking issues)

**Author:** @AllenYolk | **Base:** master | **Files:** 7 changed | **Tests:** 23/23 passing

---

### ✅ Looks Good

**Architecture & Design**
- Cleanly replaces the dual-tracking system (`ActiveModuleTracker` + `SpikeSimEventTraceMode`) with the centralized `DispatchCounterMode`, reducing coupling and reusing shared infrastructure. This is the right direction.
- Tight scope to `nn.Conv2d` forward inference only via `_is_forward_inference_conv()` — excludes training-mode convs, transposed convs, grouped convs, and non-Conv2d `aten.convolution.default` calls.
- Dense PE-cycle accounting as the default `activity_mode` correctly aligns with the original SpikeSim methodology; event mode is appropriately demoted to an experimental opt-in (`activity_mode="event"`).

**Robustness**
- Neuron type validation (`require_if_lif_neurons=True`) is well-designed: opt-out default, bilingual error messages, strict mode, and a disable flag for advanced users.
- Training guard at both `estimate_spikesim_event_energy()` (outer) and the dispatch-level `_is_forward_inference_conv()` (inner) — defense in depth.
- `torch.no_grad()` wrapper prevents accidental gradient tracking during profiling.
- Shape-mismatch detection and warning per stage is a useful diagnostic.

**Test Coverage**
- New tests cover all major scenarios: dense mode sparsity independence (`test_spikesim_default_dense_mode_ignores_input_sparsity`), training scope guard (`test_spikesim_training_scope_warns_or_raises`), IF/LIF validation (`test_spikesim_allows_if_and_lif_neurons`, `test_spikesim_unsupported_neuron_warns_or_raises`, `test_spikesim_unsupported_neuron_check_can_be_disabled`).
- All 23 existing and new tests pass.

**Cleanup**
- Removes the deleted `trace.py` (replaced by `DispatchCounterMode`) and the dead `_is_spike_like` / `_compute_spikesim_dense_stage_energy` functions. Good hygiene.

---

### 💡 Suggestions

See inline comments for:

1. **`config.py:130`** — Dead code: `pe_cycle_energy_pj` property is defined but never used. The energy pipeline uses `pe_cycle_energy_for_kernel_pj()`. Either document it as a convenience API for 3×3 or remove it.

2. **`counter.py:393`** — Redundant double computation: `_dense_pe_cycles()` is called twice per convolution dispatch (once in `_update_stage`, once in `_handle_convolution` return). Works correctly, but could be refactored to avoid the duplicate work.

Additional design notes:

3. **`core.py:246-249`** — `counts["dense_pe_cycle_count"]` derives from `self._counter.get_total()` which sums *all* records under `"Global"`. Currently only `aten.convolution.default` has a rule, so this is correct as a proxy. But it's a fragile invariant — if someone adds another rule later, `dense_pe_cycle_count` silently sums across ops. Consider using `records["Global"].get(aten.convolution.default, 0)` instead for explicitness.

4. **`core.py:400-401`** — Mutating `report.warnings.insert(0, message)` after the report is constructed is functional but slightly unusual. Consider passing the training warning into `get_report()` via the existing `add_warnings()` mechanism instead.

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Reviewed 💬** (0 issues, 2 suggestions)

**PR:** #682 — Align the SpikeSim dense energy consumption model and tighten the runtime scope
**Author:** @AllenYolk
**Files changed:** 7 (+368 −209)

---

### ✅ What Was Done Well

- **Clean architectural shift from dual-tracker to single DispatchCounterMode.** The old `SpikeSimEventTraceMode` + `ActiveModuleTracker` pair is replaced by a single `DispatchCounterMode` from `base.py`, eliminating the manual scope-injection hack (`__spikesim_scope__` kwargs). The new approach leverages the existing `ModuleTracker` infrastructure, reducing code duplication and maintenance surface.

- **Precise scope tightening.** Stripping `aten.add.Tensor`, `aten.mm`, `aten.addmm`, `aten.bmm`, `aten.baddbmm`, and `aten.cat` rules correctly scopes SpikeSim to only `aten.convolution.default`. The new `_is_forward_inference_conv()` guard (`nn.Conv2d` + eval + no_grad) is the right semantic check — it naturally handles backward/training convs without needing separate mode detection.

- **Dense mode formula alignment.** The new `_dense_pe_cycles()` × `pe_cycle_energy_for_kernel_pj()` decomposition exactly matches the removed `_compute_spikesim_dense_stage_energy()` formula, preserving backward compatibility with the original SpikeSim dense accounting. The tests (`test_spikesim_event_energy_dense_equivalence`, `test_spikesim_event_energy_dense_xbar_component_equivalence`) verify this end-to-end.

- **Neuron validation layer.** `_unsupported_neuron_modules()` + `require_if_lif_neurons` config flag provides a clean, opt-out gating mechanism. The tuple-based dispatch on `_SUPPORTED_SPIKESIM_NEURONS` (`IFNode`, `LIFNode`, `SimpleIFNode`, `SimpleLIFNode`) is easy to extend.

- **Defensive patterns.** The `config.copy()` + independent `validate()` at both the convenience function and profiler level ensures no config mutation leaks. The `_warning_keys` dedup set prevents repetitive warnings on shared modules. The `activity_mode != "event"` guard in `compute_spikesim_event_energy_breakdown()` provides a safe fallback if called directly.

- **Test quality.** The new tests (`test_spikesim_default_dense_mode_ignores_input_sparsity`, `test_spikesim_training_scope_warns_or_raises`, `test_spikesim_unsupported_neuron_warns_or_raises`, `test_spikesim_unsupported_neuron_check_can_be_disabled`) cover the key behavioral changes with clear expected values. The reuse-scope accumulation test (`test_spikesim_event_energy_repeated_same_scope_accumulates`) correctly verifies that the removed merge-op warnings no longer fire.

---

### 💡 Suggestions

#### S1 — `config.copy()` double-copy in the hot path (Maintainability)

In `estimate_spikesim_event_energy()`:
```python
cfg = (config or SpikeSimEnergyConfig()).copy()  # copy #1
cfg.validate()
...
with SpikeSimEventEnergyProfiler(config=cfg, ...) as profiler:
```

`SpikeSimEventEnergyProfiler.__init__` then does:
```python
self.config = (config or SpikeSimEnergyConfig()).copy()  # copy #2
self.config.validate()
```

When called via `estimate_spikesim_event_energy`, the config is copied and validated twice. This is harmless (dataclass copy is cheap) and provides defense-in-depth, but the second copy in the profiler could be skipped when `config` is already a validated copy. Consider adding a private parameter like `_skip_validation: bool = False` to the profiler, or simply documenting the double-copy as intentional.

#### S2 — Empty `event_stats_by_stage` for dense mode (Documentation)

When `activity_mode="dense"`, `event_stats_by_stage` in the report still contains the stage-level data (including `active_patch_tile_count`, `active_row_count`, etc.) even though these equal the `dense_*` values. This is mathematically correct (the stats track both active and dense), but may confuse users who see "event" stats in a dense-mode report. Consider adding a note in the `SpikeSimEventEnergyReport` docstring explaining that `event_stats_by_stage` is populated regardless of mode and equals the dense baseline unless `activity_mode="event"` is set.

---

### 📝 Notes

- The class names `SpikeSimEventEnergyReport` and `SpikeSimEventEnergyProfiler` retain "Event" in their names despite now supporting both modes. This is reasonable for backward compatibility, but a future rename to `SpikeSimEnergyReport`/`SpikeSimEnergyProfiler` would reduce confusion once the API stabilizes.
- The `_dense_event_counts()` call in `_update_stage()` is unconditional (always computed), which is correct because the dense stats serve as the baseline accumulator regardless of activity mode.

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: Reviewed 💬** (3 suggestions)

**PR:** #682 — Align the SpikeSim dense energy consumption model and tighten the runtime scope
**Author:** @AllenYolk
**Files changed:** 7 (+409 −240)

---

### 👍 Looks Good

- **Clean separation of concerns:** The new `compute_spikesim_dense_energy_breakdown` / `compute_spikesim_event_energy_breakdown` split in `formulas.py` is clean. The old `_compute_spikesim_dense_stage_energy` was a private function inside `formulas.py` — extracting it into a proper public function is an improvement.

- **Correct dense energy formula:** The new `_dense_pe_cycles(w, out)` returns `ceil(in_c / xbar_size) * ceil(out_c / xbar_size) * num_sites`, and `pe_cycle_energy_for_kernel_pj(kernel_size)` returns `patch_control + neuron + (xbar_size / 8) * k_h * k_w * xbar_array_energy`. This is mathematically equivalent to the old `_compute_spikesim_dense_stage_energy`, confirming the alignment claim.

- **Eliminated noise from unrelated op warnings:** Removing the merge/linear/concat counting rules from `SpikeSimCounter` is the right call. The old `SpikeSimEventCounter` would warn on `aten.add.Tensor`, `aten.mm.default`, etc., which are outside SpikeSim's scope. Now these ops are silently ignored (handled by `DispatchCounterMode`'s `has_rule` check), and the test `test_spikesim_event_energy_scalar_add_does_not_warn` confirms this.

- **Clean dispatch architecture:** Replacing the two-layer `ActiveModuleTracker` + `SpikeSimEventTraceMode` with a single `DispatchCounterMode` aligns the SpikeSim profiler with the rest of `op_counter` and removes the redundant custom `SpikeSimEventTraceMode` (deleted `trace.py`).

- **Good test coverage:** 23 tests passing, covering dense equivalence, event-driven counting, multi-step, training scope, neuron validation, grouped/transposed conv, and edge cases. The new tests (`test_spikesim_default_dense_mode_ignores_input_sparsity`, `test_spikesim_training_scope_warns_or_raises`, etc.) cover the new features well.

---

### 💡 Suggestions (Documentation)

**1. `SpikeSimEnergyConfig` docstring still says "event-driven" (`config.py:33`)**

The config now supports both `activity_mode="dense"` (default) and `activity_mode="event"`, but the docstring still describes it as:

> "Runtime energy configuration for the event-driven SpikeSim estimator."

Since `activity_mode="dense"` is now the default and the config is used by both modes, this should be updated to something like:

> "Runtime energy configuration for the SpikeSim-aligned energy estimator."

**2. `SpikeSimEnergyReport` docstring still says "event-driven" (`core.py:72, 82`)**

Both the Chinese and English sections say "Report for the event-driven SpikeSim runtime energy estimator." Since the report is now generated for both dense and event modes, "event-driven" is misleading.

**3. `breakdown_pj` and `counts` fields in `SpikeSimEnergyReport` (`core.py:94-95`)**

The new `breakdown_pj: dict[str, float]` and `counts: dict[str, int]` fields are added to the report dataclass but their semantics aren't explained in the docstring. A brief note in the class docstring would help users reading the report output — e.g., that `breakdown_pj` contains total component-wise energy (by component, summed across stages) and `counts` contains `dense_pe_cycle_count` and `stage_count`.

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Reviewed 💬** (0 critical, 0 warnings, 3 suggestions)

**PR:** #682 — Align the SpikeSim dense energy consumption model and tighten the runtime scope
**Author:** @AllenYolk
**Files changed:** 7 (+412 −243)

---

### 💡 Suggestions

- **(Documentation) `formulas.py` — `compute_spikesim_event_energy_breakdown` silently delegates to `compute_spikesim_dense_energy_breakdown` when `activity_mode != "event"`.** The function name "event_energy" is misleading when running in dense mode. Consider: (a) adding a brief note in the docstring documenting this fallback behavior, or (b) renaming the helper to clarify it serves both modes.

- **(Documentation) `counter.py:410-416` (`_is_forward_inference_conv`) — `torch.is_grad_enabled()` is global state, not module-specific.** When users use `SpikeSimEnergyProfiler` directly (not via `estimate_spikesim_event_energy`) and call `model.eval()` but forget `torch.no_grad()`, they'll get an opaque "outside-scope" warning instead of a clear message. The convenience function handles this correctly, but a docstring note on the profiler or counter would help direct users to the right pattern.

- **(Maintainability) `counter.py:158-167` (`_dense_pe_cycles`) — This method only accesses `self.config`.** Consider making it a `@staticmethod` accepting `config` as a parameter, or a module-level function, to make the data flow explicit and reduce coupling.

### ✅ Looks Good

- **Clean architecture** — Good separation of concerns: `config` (data), `counter` (aten-level dispatch), `formulas` (energy computation), `core` (profiler orchestration).
- **Eliminated duplication** — Replacing custom `SpikeSimEventTraceMode` with the shared `DispatchCounterMode` from `base.py` removes the duplicated `TorchDispatchMode` subclass and its manual `__torch_dispatch__` boilerplate.
- **Backward compatibility** — `SpikeSimEventEnergyReport = SpikeSimEnergyReport` and `SpikeSimEventEnergyProfiler = SpikeSimEnergyProfiler` aliases at both module and class level ensure existing users don't break.
- **Neuron validation** — `require_if_lif_neurons` with clear warnings/errors is a thoughtful safeguard against misusing SpikeSim hardware models with incompatible neuron types.
- **Dense formula accuracy** — Verified that `_dense_pe_cycles(p_i * q_i * num_sites)` and `pe_cycle_energy_for_kernel_pj` match the deleted `_compute_spikesim_dense_stage_energy` exactly.
- **Test coverage** — New tests cover default dense mode (sparsity-independence), training scope validation, IF/LIF neuron acceptance, unsupported neuron rejection, and disabling the neuron check. Existing "event" tests correctly set `activity_mode="event"` to maintain coverage.
- **No double-counting** — `_exclude_python_dispatch_guard()` correctly prevents recursive counting from the internal `F.conv2d` call in `_spike_event_counts`.

---
*Reviewed by Hermes Agent*

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 4]
```diff
@@ -115,11 +104,15 @@ def count(
...
-        return value
+        return int(
+            self.rules[func](
+                args,
+                kwargs,
+                out,
+                active_modules=active_modules,
+                parent_names=parent_names,
+            )
+        )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/spikesim/counter.py:115]: not needed. record() is called by dispatch counter mode

File: spikingjelly/activation_based/op_counter/spikesim/counter.py [hunk 5]
```diff
@@ -360,48 +383,37 @@ def _handle_convolution(self, scope: str, args: tuple[Any, ...], out) -> None:
...
-            f"SpikeSim event energy ignores concat-like op {op_name}.",
-        )
-        return 0
+    def _is_forward_inference_conv(self, active_modules: set[nn.Module] | None) -> bool:
+        modules = active_modules or set()
+        conv2d_modules = [module for module in modules if isinstance(module, nn.Conv2d)]
+        if len(conv2d_modules) != 1:
+            return False
+        conv = conv2d_modules[0]
+        return (not conv.training) and (not torch.is_grad_enabled())
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/op_counter/spikesim/counter.py:416]: len(conv2d_modules) != 1 is necessary for spikesim-style energy estimation

**Comment by AllenYolk** [comment]
@coderabbitai review
@gemini-code-assist review

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: Approved ✅** (0 issues, 1 suggestion)

**PR:** #682 — Align the SpikeSim dense energy consumption model and tighten the runtime scope
**Author:** @AllenYolk
**Files changed:** 7 (+428 −246)

---

### ✅ What's Been Done Well

**Architecture simplification.** Eliminating the custom `SpikeSimEventTraceMode` and delegating to `DispatchCounterMode` is the right move — it eliminates code duplication with the rest of `op_counter`, removes the manual `__torch_dispatch__` lifecycle management, and correctly leverages `ModuleTracker`'s parent-scope recording. The deleted `trace.py` (37 lines) is a net positive.

**Dense mode as default.** Setting `activity_mode="dense"` as the default correctly aligns with the original SpikeSim dense PE-cycle accounting model. The `_dense_pe_cycles` formula (`ceil(in_c / xbar) x ceil(out_c / xbar) x num_sites`) is mathematically equivalent to the deleted `_compute_spikesim_dense_stage_energy`'s cycle count, and `pe_cycle_energy_for_kernel_pj()` cleanly captures per-cycle energy as a function of kernel size — fixing the previously hardcoded 3x3 assumption.

**Scope tightening.** Removal of non-conv rules and addition of `_is_forward_inference_conv` bring the profiler's scope in line with its documented purpose (Conv2d forward inference only). The `nn.Conv2d` isinstance check correctly handles `layer.Conv2d` (inherits from `nn.Conv2d`).

**Neuron validation.** The `require_if_lif_neurons` flag and `_unsupported_neuron_modules` helper prevent applying the SpikeSim model to incompatible neurons, with a clean warn/raise path.

**Backwards compatibility.** `SpikeSimEventEnergyReport = SpikeSimEnergyReport` and `SpikeSimEventEnergyProfiler = SpikeSimEnergyProfiler` as type aliases preserve existing imports and isinstance checks. The `estimate_spikesim_event_energy` signature is unchanged.

**Test coverage.** 5 new tests cover: dense mode input-sparsity invariance, training mode, IF/LIF neuron acceptance, unsupported neuron rejection, and the disable-neuron-check path. All 23 tests pass locally.

---

### 💡 Suggestion (Non-blocking)

**`training_message` constructed unconditionally** — In `estimate_spikesim_event_energy` (line 380-383), the string is allocated even when `model.training` is False. Consider moving it into the `if model.training:` block or using a module-level constant to avoid an unnecessary allocation on the common eval-mode path.

---

### Iteration Quality

The commit history shows five incremental fix rounds addressing feedback from Gemini Code Assist, CodeRabbit, and self-review — each focused on a single concern (var reuse, dead code, double compute, constant naming, input validation). This is the right way to iterate a refactoring PR.

---
*Reviewed by Hermes Agent*
